daml-foundations/daml-ghc/src/DA/Daml/GHC/Compiler/Convert.hs

```diff
@@ -336,26 +336,52 @@ archiveChoice signatories = TemplateChoice{..}
-        `App` consuming `App` Var controller `App` choice `App` _exercise) = do
-    chcConsuming <- f (10 :: Int) consuming
+data Consuming = PreConsuming
+               | NonConsuming
+               | PostConsuming
+               deriving (Eq)
+
+convertChoice :: Env -> LF.Expr -> GHC.Expr Var -> ConvertM TemplateChoice
+convertChoice env signatories
+  (VarIs "C:Choice" `App`
+     Type tmpl@(TypeCon tmplTyCon []) `App`
+       Type (TypeCon chc []) `App`
+         Type result `App`
+           _templateDict `App`
+             consuming `App`
+               Var controller `App`
+                 choice `App` _exercise) = do
+    consumption <- f (10 :: Int) consuming
+    let chcConsuming = consumption == PreConsuming -- Runtime should auto-archive?
     argType <- convertType env $ TypeCon chc []
     let chcArgBinder = (mkVar "arg", argType)
     tmplType <- convertType env tmpl
+    tmplTyCon' <- convertQualified env tmplTyCon
     chcReturnType <- convertType env result
     controllerExpr <- envFindBind env controller >>= convertExpr env
     let chcControllers = case controllerExpr of
-            -- NOTE(MH): We drop the second argument to `controllerExpr` when
-            -- it is unused. This is necessary to make sure that a
-            -- non-flexible controller expression does not mention the choice
-            -- argument `argVar`.
-            ETmLam thisBndr (ETmLam argBndr body)
-              | fst argBndr `Set.notMember` cata freeVarsStep body ->
-                ETmLam thisBndr body `ETmApp` thisVar
-            _ ->
-                controllerExpr `ETmApp` thisVar `ETmApp` argVar
-    chcUpdate <- fmap (\u -> u `ETmApp` thisVar `ETmApp` selfVar `ETmApp` argVar) (convertExpr env choice)
+          -- NOTE(MH): We drop the second argument to `controllerExpr` when
+          -- it is unused. This is necessary to make sure that a
+          -- non-flexible controller expression does not mention the choice
+          -- argument `argVar`.
+          ETmLam thisBndr (ETmLam argBndr body)
+            | fst argBndr `Set.notMember` cata freeVarsStep body ->
+              ETmLam thisBndr body `ETmApp` thisVar
+          _ -> controllerExpr `ETmApp` thisVar `ETmApp` argVar
+    expr <- fmap (\u -> u `ETmApp` thisVar `ETmApp` selfVar `ETmApp` argVar) (convertExpr env choice)
+    let chcUpdate =
+          if consumption /= PostConsuming then expr
+          else
+            -- NOTE(SF): Support for 'postconsuming' choices. The idea
+            -- is to evaluate the user provided choice body and
+            -- following that, archive. That is, in pseduo-code, we are
+            -- going for an expression like this:
+            --     expr this self arg >>= \res ->
+            --     archive signatories self >>= \_ -> return res
+            let archive = EUpdate $ UExercise tmplTyCon' (mkChoiceName "Archive") selfVar signatories mkEUnit
+            in EUpdate $ UBind (Binding (mkVar "res", chcReturnType) expr) $
```

Comment:
If the EUpdate's aligned beneath each other i would find it more readadble

Reply:
✔️